dts/bindings/rtc/nordic,nrf-rtc.yaml

```diff
@@ -23,6 +23,20 @@ properties:
       generation: define
       category: required
 
+    # Guard period should be set to the maximal latency that can occur when
+    # controlling the counter (e.g. for how long operation may be
+    # interrupted). Guard period limits maximal alarm value that can be set
+    # to <top value> - <guard period in ticks> from now. Limitation applies
+    # to relative and absolute alarms. Longer alarm values will be
+    # interpretted as set too late and will expire immediately. If instance
+    # is used only from interrupt context value probably can be decreased.
+    # However, if used from low priority thread it may need to be increased.
+    guard-period:
```

Comment:
I'm not sure what this is trying to describe.

RTC requires that a value stored in a `CC` register be at least 2 ticks  past the current `COUNTER` to guarantee a `COMPARE` event fires.  Is it that `2` that's supposed to be captured here?  I'll assume so, though for some reason the value in the DTS files is `10`.

That limitation has no relation to interrupt priority, so much of the description brings in irrelevant concepts.  Also `guard-period` does not suggest what this value is for in terms related to RTC or counter.  Something like `compare-tick-offset` would be closer, though I don't love it.

Further, whatever this property describes probably applies to other SOC asynchronous timer infrastructure.  The property should be described in the parent `rtc.yaml`.  Something like it probably needs to be in a parent `timer.yaml` that does not currently exist.

The description field should be useful, like the other ones in the parent `rtc.yaml`.

Reply:
Value that should be set to guard-period is the maximal potential preemption period that can occur during setting the alarm. If alarm is only controlled from interrupt context then guard period can be small, if from any context then it must be increased. Increasing guard-period reduced maximal alarm value that can be set (e.g in RTC case 2^24-guard-period-in-ticks) so if user is unsure it should be increased. By default i set it to 10ms which i thought should be ok for rtos system.

I agree that it's difficult to describe it well. I will try to look into improving this description.

Reply:
No.

The "maximal potential preemption period" is necessarily dependent on what the application does.  Maybe I've got an app that has a critical data collection that takes 100 ms and cannot be interrupted.

This is not a hardware-related property, and does not belong in the device tree (unlike the 2 tick counter offset, which does).  If you have to have it to make the algorithm work, it should be a property passed to the counter when the application configures it, just as with top value.

And wherever it goes, you should specify the units of the value; I had no idea it wasn't RTC ticks.  Milliseconds for time resolution is a Zephyr design decision  unrelated to Nordic hardware; it's also one that will very soon have to be reconsidered because it's a poor choice given current MCU operating frequencies.

Reply:
> This is not a hardware-related property, and does not belong in the device tree

I think there is an overall assumption in zephyr that driver instance configuration is part of device tree. Whether it's hw related or not. I initially was against but accepted as imo it's better to have instance configuration in one place rather than split across two configuration tools.

Reply:
>Milliseconds for time resolution is a Zephyr design decision unrelated to Nordic hardware; it's also one that will very soon have to be reconsidered because it's a poor choice given current MCU operating frequencies.

I think that milliseconds is more user friendly since it is still rough value but I can change it to ticks if you insist and even set the default to be much bigger (like 32768). That limits RTC range to 511s instead of 512s (assuming 0 prescaler and no change in top value) which will be unnoticed in majority of cases.

I need to decrease it for the tests where i use smaller top value.

> it should be a property passed to the counter when the application configures it, just as with top value.

Would you see that as part of the API, something like: `set_top_value(dev, ... , guard)`?
@pizi-nordic @anangl opinions? @pabigot pros, cons?

Reply:
> > This is not a hardware-related property, and does not belong in the device tree
> 
> I think there is an overall assumption in zephyr that driver instance configuration is part of device tree. Whether it's hw related or not. I initially was against but accepted as imo it's better to have instance configuration in one place rather than split across two configuration tools.

That is certainly not the impression I had, based on the expectation that device-tree content was to be OS-independent (shared with other projects).  @galak ?